File: src/unit/content/style.c
#include <hb/proc.h>

static void _parse_comment(hb_proc* proc)
{
	hb_proc_require_match(proc, "/*");

	// Unlike script tags, style comments do NOT end at closing tag.
	while (!hb_proc_accept_if_matches(proc, "*/")) {
		hb_proc_accept(proc);
	}
}

static void _parse_string(hb_proc* proc)
{
	hb_rune delim = hb_proc_accept(proc);

	if (delim != '"' && delim != '\'') {
		hb_proc_error(proc, HB_ERR_PARSE_EXPECTED_NOT_FOUND,
			      "Expected CSS string delimiter");
	}

	bool escaping = false;

	while (true) {
		hb_rune c = hb_proc_accept(proc);

		if (c == '\\') {
			escaping = !escaping;
			continue;
		}

		if (c == delim && !escaping) {
			break;
		}

		if (hb_proc_accept_if_matches_line_terminator(proc)) {
			if (!escaping) {
				hb_proc_error(proc,
					      HB_ERR_PARSE_EXPECTED_NOT_FOUND,
					      "Unterminated CSS string");
			}
		}

		escaping = false;
	}
}

void hb_unit_content_style(hb_proc* proc)
{
	while (!hb_proc_matches(proc, "</")) {
		if (hb_proc_matches(proc, "/*")) {
			_parse_comment(proc);
		} else {
			switch (hb_proc_peek(proc)) {
			case '\'':
			case '"':
				_parse_string(proc);
				break;
			default:
				hb_proc_accept(proc);
			}
		}
	}
}

File: src/proc/style.rs
use crate::proc::Processor;
use crate::err::{HbRes, HbErr};
use crate::code::Code;

fn is_string_delimiter(c: u8) -> bool {
    match c {
        b'"' | b'\'' => true,
        _ => false,
    }
}

fn parse_comment<D: Code>(proc: &Processor<D>) -> HbRes<()> {
    proc.match_seq(b"/*").expect().keep();

    // Unlike script tags, style comments do NOT end at closing tag.
    while !proc.match_seq(b"*/").keep().matched() {
        proc.accept();
    };

    Ok(())
}

fn parse_string<D: Code>(proc: &Processor<D>) -> HbRes<()> {
    let delim = proc.match_pred(is_string_delimiter).expect().keep().char();

    let mut escaping = false;

    loop {
        let c = proc.accept()?;

        if c == b'\\' {
            escaping = !escaping;
            continue;
        }

        if c == delim && !escaping {
            break;
        }

        if proc.match_line_terminator().keep().matched() {
            if !escaping {
                // TODO Use better error type.
                return Err(HbErr::ExpectedNotFound("Unterminated CSS string"));
            }
        }

        escaping = false;
    };

    Ok(())
}

pub fn process_style<D: Code>(proc: &Processor<D>) -> HbRes<()> {
    while !proc.match_seq(b"</").matched() {
        if proc.match_seq(b"/*").matched() {
            parse_comment(proc)?;
        } else if proc.match_pred(is_string_delimiter).matched() {
            parse_string(proc)?;
        } else {
            proc.accept()?;
        }
    };

    Ok(())
}
